apis/Google.Cloud.EntityFrameworkCore.Spanner/Google.Cloud.EntityFrameworkCore.Spanner/Extensions/SpannerServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Cloud.EntityFrameworkCore.Spanner.Infrastructure.Internal;
+using Google.Cloud.EntityFrameworkCore.Spanner.Metadata.Conventions;
+using Google.Cloud.EntityFrameworkCore.Spanner.Query.ExpressionTranslators.Internal;
+using Google.Cloud.EntityFrameworkCore.Spanner.Query.Internal;
+using Google.Cloud.EntityFrameworkCore.Spanner.Query.Sql.Internal;
+using Google.Cloud.EntityFrameworkCore.Spanner.Storage.Internal;
+using Google.Cloud.EntityFrameworkCore.Spanner.Update.Internal;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal;
+using Microsoft.EntityFrameworkCore.Query;
+using Microsoft.EntityFrameworkCore.Query.ExpressionTranslators;
+using Microsoft.EntityFrameworkCore.Query.Sql;
+using Microsoft.EntityFrameworkCore.Storage;
+using Microsoft.EntityFrameworkCore.Update;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Spanner specific extension methods for <see cref="IServiceCollection" />.
+    /// </summary>
+    public static class SpannerServiceCollectionExtensions
+    {
+        /// <summary>
+        /// </summary>
+        /// <param name="serviceCollection"></param>
+        /// <returns></returns>
+        public static IServiceCollection AddEntityFrameworkSpanner(this IServiceCollection serviceCollection)
+        {
+            GaxPreconditions.CheckNotNull(serviceCollection, nameof(serviceCollection));
+
+            var builder = new EntityFrameworkRelationalServicesBuilder(serviceCollection)
```

Comment:
yeah its kind of an odd pattern that it takes in the collection that gets modified.
This method is actually the single most important method that exposes all of our services to EF.  You'll see I left some there commented out because it's not quite trivial to figure out which services we actually need.